parser/impl/parser.cpp

```diff
@@ -7,51 +7,289 @@ namespace parser
+            }
+        }
+        return std::make_shared<ast::Program>(programNode);
+    }
+
+    sPtr<ast::Routine> Parser::parseRoutine() {
+        ast::Routine routineNode;
+        lexer::Token currentToken = m_lexer.Next();
+        if (currentToken.type != lexer::TokenType::Routine) {
+            m_errors.push_back(Error{
+                .pos = currentToken.pos,
+                .message = "Expected 'routine' keyword but did not find one",
+            });
+            return nullptr;
+        }
+        currentToken = skipWhile(isNewLine);
+
+        if (currentToken.type != lexer::TokenType::Identifier) {
+            m_errors.push_back(Error{
+                .pos = currentToken.pos,
+                .message = "Expected an identifier but did not find one",
+            });
+            // skip till "end" keyword
+            while(m_lexer.Next().type != lexer::TokenType::End);
+            return nullptr;
+        }
+        routineNode.name = currentToken;
+
+        currentToken = skipWhile(isNewLine);
+
+        if (currentToken.type != lexer::TokenType::OpenParen) {
+            m_errors.push_back(Error{
+                .pos = currentToken.pos,
+                .message = "Expected to find '('",
+            });
+            // skip till "end" keyword
+            while(m_lexer.Next().type != lexer::TokenType::End);
+            return nullptr;
+        }
+
+        while(currentToken.type != lexer::TokenType::CloseParen) {
+            routineNode.parameters.push_back(parseParameter());
+            currentToken = m_lexer.Next();
+            if (currentToken.type == lexer::TokenType::Comma) {
+                currentToken = m_lexer.Next();
+            }
+        }
+        currentToken = skipWhile(isNewLine);
+        if (currentToken.type != lexer::TokenType::Colon && currentToken.type != lexer::TokenType::Is) {
+            m_errors.push_back(Error{
+                .pos = currentToken.pos,
+                .message = "Unexpected token",
+            });
+            // skip till "end" keyword
+            while(m_lexer.Next().type != lexer::TokenType::End);
+            return nullptr;
+        }
+        if (currentToken.type == lexer::TokenType::Colon) {
+            currentToken = skipWhile(isNewLine);
+            routineNode.returnType = parseType();
+        }
+        currentToken = skipWhile(isNewLine);
+
+        if (currentToken.type != lexer::TokenType::Is) {
+            m_errors.push_back(Error{
+                .pos = currentToken.pos,
+                .message = "Expected to find keyword 'is'",
+            });
+            // skip till "end" keyword
+            while(m_lexer.Next().type != lexer::TokenType::End);
+            return nullptr;
+        }
+        currentToken = skipWhile(isNewLine);
+        routineNode.body = parseBody();
+        if (currentToken.type != lexer::TokenType::End) {
+            m_errors.push_back(Error{
```

Comment:
Isn't this part of code duplicated? Can we make some function, which take error msg, token `pos` and:
1. shows error
2. skips till the end of file (probably do it conditionally (3rd parameter))

Reply:
Right, but a function that just wraps the call to `m_errors.push_back` is not that much more helpful since the Error still has to be customized for every instance. The skipping to the end (or to some token) can be useful, but I want to do it in a neat way that would probably be made clearer after more examples (I'm especially waiting for @kureduro to push his code because I think he may have already implemented something like that)